service/sharddistributor/executorclient/clientimpl.go

```diff
@@ -0,0 +1,122 @@
+}
+
+func (e *executorImpl[SP]) Start(ctx context.Context) {
+	go e.heartbeatloop(ctx)
+}
+
+func (e *executorImpl[SP]) Stop() {
+	close(e.stopC)
+}
+
+func (e *executorImpl[SP]) GetShardProcess(shardID string) (SP, error) {
+	shardProcess, ok := e.shardProcessors[shardID]
+	if !ok {
+		var zero SP
+		return zero, fmt.Errorf("shard process not found for shard ID: %s", shardID)
+	}
+	return shardProcess, nil
+}
+
+func (e *executorImpl[SP]) heartbeatloop(ctx context.Context) {
+	heartBeatTicker := e.timeSource.NewTicker(e.heartBeatInterval)
+	defer heartBeatTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+		case <-e.stopC:
+			e.logger.Info("shard distributorexecutor stopped")
+			e.stopShardProcessors()
+			return
+		case <-heartBeatTicker.Chan():
+			shardAssignment, err := e.heartbeat(ctx)
+			if err != nil {
+				e.logger.Error("failed to heartbeat", tag.Error(err))
+				continue // TODO: should we stop the executor, and drop all the shards?
+			}
+			e.updateShardAssignment(ctx, shardAssignment)
+		}
+	}
+}
+
+func (e *executorImpl[SP]) heartbeat(ctx context.Context) (shardAssignments map[string]*types.ShardAssignment, err error) {
+	// Fill in the shard status reports
+	shardStatusReports := make(map[string]*types.ShardStatusReport)
+	for shardID, processor := range e.shardProcessors {
+		shardStatusReports[shardID] = &types.ShardStatusReport{
+			ShardLoad: processor.GetShardLoad(),
+			Status:    types.ShardStatusREADY,
+		}
+	}
+
+	// Create the request
+	request := &types.ExecutorHeartbeatRequest{
+		Namespace:          e.namespace,
+		ExecutorID:         e.executorID,
+		Status:             types.ExecutorStatusACTIVE,
+		ShardStatusReports: shardStatusReports,
+	}
+
+	// Send the request
+	response, err := e.shardDistributorClient.Heartbeat(ctx, request)
+	if err != nil {
+		return nil, fmt.Errorf("send heartbeat: %w", err)
+	}
+
+	return response.ShardAssignments, nil
+}
+
+func (e *executorImpl[SP]) updateShardAssignment(ctx context.Context, shardAssignments map[string]*types.ShardAssignment) {
+	// Stop shard processing for shards not assigned to this executor
+	for shardID, processor := range e.shardProcessors {
+		if assignment, ok := shardAssignments[shardID]; !ok || assignment.Status != types.AssignmentStatusREADY {
+			processor.Stop()
+			delete(e.shardProcessors, shardID)
+		}
+	}
+
+	// Start shard processing for shards assigned to this executor
+	for shardID, assignment := range shardAssignments {
+		if assignment.Status == types.AssignmentStatusREADY {
+			if _, ok := e.shardProcessors[shardID]; !ok {
+				processor, err := e.shardProcessorFactory.NewShardProcessor(shardID)
+				if err != nil {
+					e.logger.Error("failed to create shard processor", tag.Error(err))
+					continue
+				}
+				processor.Start(ctx)
+				e.shardProcessors[shardID] = processor
+			}
+		}
+	}
+}
+
+func (e *executorImpl[SP]) stopShardProcessors() {
+	for _, processor := range e.shardProcessors {
```

Comment:
The same as updateShardAssignment. Stop can do a cleanup - so can take time.

Reply:
Added a go-routine for each processor starting/stopping